homeassistant/components/livisi/coordinator.py

```diff
@@ -0,0 +1,139 @@
+
+    async def _async_update_data(self):
+        """Set device list."""
+        shc_devices: dict[Any, Any] = await self.async_get_devices()
+        for device in shc_devices:
+            if device in self.devices:
+                continue
+            if device["type"] == PSS_DEVICE_TYPE:
+                if device not in self.switch_devices:
+                    self.switch_devices.append(device)
+                    async_dispatcher_send(self.hass, LIVISI_DISCOVERY_NEW, device)
+            self.devices.append(device)
+
+    async def async_setup(self) -> None:
+        """Set up the Livisi Smart Home Controller."""
+        if bool(self.aiolivisi.livisi_connection_data) is False:
+            livisi_connection_data = {
+                "ip_address": self.config_entry.data["host"],
+                "password": self.config_entry.data["password"],
+            }
+
+            await self.aiolivisi.async_set_token(
+                livisi_connection_data=livisi_connection_data
+            )
+        controller_info = await self.aiolivisi.async_get_controller()
+
+        controller_data = controller_info
+        if controller_data.get("controllerType") == "Avatar":
+            self.port = AVATAR_PORT
+            self.is_avatar = True
+        else:
+            self.port = CLASSIC_PORT
+            self.is_avatar = False
+        self.serial_number = controller_data.get("serialNumber")
+        self.controller_type = controller_data.get("controllerType")
+        self.os_version = controller_data.get("osVersion")
+
+    async def async_get_devices(self) -> list:
+        """Set the discovered devices list."""
+        shc_devices = await self.aiolivisi.async_get_devices()
+        if bool(shc_devices) is False:
+            shc_devices = await self.aiolivisi.async_get_devices()
+        return shc_devices
+
+    async def async_get_pss_state(self, capability):
+        """Set the PSS state."""
+        response = await self.aiolivisi.async_get_pss_state(capability[1:])
+        if response is None:
+            return
+        on_state = response.get("onState")
+        return on_state.get("value")
+
+    async def async_set_all_rooms(self):
+        """Set the room list."""
+        response = await self.aiolivisi.async_get_all_rooms()
+
+        for available_room in response:
+            self.rooms[available_room.get("id")] = available_room.get("config").get(
+                "name"
+            )
+
+    def on_data(self, event_data: LivisiEvent) -> None:
+        """Define a handler to fire when the data is received."""
+        source = event_data.source
+        if event_data.onState is not None:
+            device_id_state: dict = {
+                "id": source,
+                "state": event_data.onState,
+            }
+            self.async_set_updated_data(device_id_state)
+        if event_data.isReachable is not None:
+            device_id_reachability: dict = {
+                "id": source.replace("/device/", ""),
```

Comment:
We shouldn't need to do this kind of replacement. This is a device protocol detail that belongs in the device library.

Reply:
I moved this replacement in the device library, thank you.